### ISSUE TITLE
Async client does listener does not fire onFailure() when subscribe fails (returns QOS 128)

### ISSUE THREAD

**Issue by ghost** [eclipse-paho/paho.mqtt.java#216]
Bug exists in Snapshot Version 0.4.1 (Develop Branch)

Sync client properly responds to a failed subscribe (e.g. due to topic ACL) and recognizes a QOS 128 by throwing an MqttException.

Async client success callback fires in the same situation.  Wire message delivered is: (MqttActionListener.onSuccess) op: subscribe, wire: SUBACK msgId 1 granted Qos 128

Shouldn't the async client fire an onFailure() as well?

**Comment by jpwsutton** [comment]
Confirmed locally using the interop test broker:

```
Sync Client:
Sync Subscribing...
MqttException (128)
    at org.eclipse.paho.client.mqttv3.MqttClient.subscribe(MqttClient.java:329)
    at org.eclipse.paho.client.mqttv3.MqttClient.subscribe(MqttClient.java:315)
    at issue_216_async_incorrect_sub_error.main.main(main.java:24)
Async Client:
Success
Async Subscribing...
Success
```

**Comment by jpwsutton** [comment]
However this is more of an implementation problem with the Synchronous client, not the Async client.

The expected behaviour for the Async client is to only return onFailure if the Subscribe call actually fails. Getting a Granted QoS of 128 isn't actually a failure and is just the broker refusing to complete that particular subscription. This is because when you subscribe to a topic, what's going on under the covers is that you are sending an array of topics & QoSs to the broker for subscription (Most of the time you just send one, which is then made into an array of one). The SUBACK contains an array of granted QoSs which is what leads us to this problem.

Lets say you subscribe to three topics and only one of them 'fails' and has a granted QoS of 128. You would receive this array in the SUBACK : `[2,128,2]`. Has every subscription failed? No, but one or more have. To the application developer an exception or an onFailure would be quite bad as they would probably assume that **all** the subscriptions had failed, thus causing them to attempt to make them again (not a huge problem as duplicate subscriptions simply replace each other). If the reason for a subscription failing was because of permissions this would then cause the application to think that it had made none of it's subscriptions.

**What should happen?**
Assuming the Client receives the SUBACK, the subscription request has 'succeeded'. It's then up to the developer to check that the topic/s have succeeded in their individual subscriptions. This is simple in the Async client as all you have to do is call `getGrantedQoS` on the IMqttToken you get in the IMqttActionListener onSuccess callback.

In the Sync client this isn't possible because all the subscribe calls return void and so instead throw an exception, this is the incorrect behaviour. I don't think we should remove the existing behaviour as it's possible that some developers including yourself are relying on the exception being thrown. However I think we should add an additional subscribe call to the Sync client that returns an IMqttToken that can be then used to judge the granted QoS.

**tldr;** The bug is actually that an exception is thrown by the Sync client, not the lack of an error in the Async client. We need to add a better subscribe call to the Sync client API.

**Comment by ghost** [comment]
Thanks for the response - makes sense yeah, and inspecting each grantedQos is what I'll likely end up doing.  Since MqttActionListener is designed to be used by all the different Mqtt methods and not just subscribe, it is inappropriate to fire off an onFailure() for only partial subscription fails.  It was just made more apparent in my case since my app client only allows one subscription at a time.

**Comment by jpwsutton** [comment]
This was resolved in PR #266 by adding an additional set of subscribe calls to the MqttClient called subscribeWithResponse. These will return an IMqttToken and will not throw an exception if a QoS of 128 is returned.
